[FNet P0] add PretrainedConfig and unit test

robinbg opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->

### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->

### Description
<!-- Describe what this PR does -->

robinbg committed:
Create configuration.py

robinbg committed:
Update modeling.py

robinbg committed:
Update __init__.py

robinbg committed:
Update configuration.py

robinbg committed:
Update modeling.py

robinbg committed:
Update configuration.py

robinbg committed:
Update modeling.py

robinbg committed:
Update modeling.py

robinbg committed:
Update modeling.py

robinbg committed:
Update modeling.py

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/PaddlePaddle/PaddleNLP?pullRequest=4943) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/PaddlePaddle/PaddleNLP?pullRequest=4943) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: robinbg<br/>:x: robinbg@foxmail.com<br/><hr/>**robinbg@foxmail.com** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/PaddlePaddle/PaddleNLP?pullRequest=4943) it.</sub>

robinbg committed:
Update modeling.py

sijunhe reviewed (commented):
您好，欢迎参加飞桨黑客松。这个PR是一个好的开始，后续需要写好test_modeling和test_tokenizer单测，并且做到单测通过

robinbg commented:
请问lint通不过是什么问题呢

sijunhe commented:
> 请问lint通不过是什么问题呢

lint的作用是维持我们的代码标准。具体的报错可以点进Lint的action details就能看到。推荐您在本地配好我们的pre-commit, 这样子每次git commit都会自动检查代码风格

Commit:
test

robinbg committed:
Update modeling.py

robinbg committed:
Update modeling.py

robinbg committed:
Update configuration.py

robinbg committed:
Update modeling.py

robinbg committed:
Update modeling.py

robinbg committed:
Update configuration.py

robinbg committed:
Update configuration.py

sijunhe commented on paddlenlp/transformers/fnet/configuration.py:
```diff
@@ -0,0 +1,159 @@
+}
+
+
+class FnetConfig(PretrainedConfig):
```
`FNetConfig` 需要对齐[HuggingFace Transformers的FNetConfig](https://github.com/huggingface/transformers/blob/v4.26.1/src/transformers/models/fnet/configuration_fnet.py#L30). `use_tpu_fourier_optimizations`和`tpu_short_seq_length`这两个参数可以不用添加

sijunhe reviewed (commented):
当前的单测错误都是import时候名字不对导致的，需要修改一下 `ImportError: cannot import name 'FNetConfig' from 'paddlenlp.transformers.fnet.configuration'`

robinbg committed:
Update configuration.py

robinbg committed:
Update configuration.py

robinbg committed:
Merge branch 'PaddlePaddle:develop' into robinbg-add-FNet-config

robinbg committed:
Update modeling.py

robinbg commented:
您好，那个CLA的邮箱也是我的（事实上，那个就是我登录时用的邮箱），然后我通过vscode commit过几次代码，我只有这一个GitHub号，请问现在我应该怎么做？

sijunhe commented:
> 您好，那个CLA的邮箱也是我的（事实上，那个就是我登录时用的邮箱），然后我通过vscode commit过几次代码，我只有这一个GitHub号，请问现在我应该怎么做？

原因是你中间有一次commit用了别的邮箱，这个PR就算了，下次注意一直用现有的github账号哈
![image](https://user-images.githubusercontent.com/11987277/221087461-477497b5-6ca9-4165-8e0c-22c1250f8b00.png)

sijunhe commented:
请增加test_modeling和test_tokenizer单测

robinbg committed:
Create __init__.py

robinbg committed:
Create test_modeling.py

robinbg committed:
Create test_tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   tests/transformers/fnet/test_modeling.py
	modified:   tests/transformers/fnet/test_tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   tests/transformers/fnet/__init__.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   tests/transformers/fnet/test_modeling.py
	modified:   tests/transformers/fnet/test_tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   tests/transformers/fnet/test_modeling.py
	modified:   tests/transformers/fnet/test_tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   tests/transformers/fnet/test_modeling.py

robinbg committed:
On branch robinbg-add-FNet-config

sijunhe commented on paddlenlp/transformers/fnet/modeling.py:
```diff
@@ -287,46 +295,96 @@ class FNetPretrainedModel(PretrainedModel):
+    pretrained_resource_files_map = FNET_PRETRAINED_RESOURCE_FILES_MAP
     base_model_prefix = "fnet"
 
+    @classmethod
+    def _get_name_mappings(cls, config: FNetConfig) -> List[StateDictNameMapping]:
```
这一块不属于PR范畴内，不需要添加

robinbg replied:
ok

sijunhe reviewed (commented):
`fnet/modeling.py`内需要适配config, 即大部分的类的constructor的输入参数都应该是config, 不是一个个的参数

robinbg committed:
Update test_tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config

robinbg committed:
Update test_tokenizer.py

robinbg committed:
Update test_tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
	modified:   .copyright.hook

robinbg committed:
Update test_tokenizer.py

robinbg committed:
Update modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py
	modified:   tests/transformers/fnet/test_modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py
	modified:   tests/transformers/fnet/test_modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py
	modified:   tests/transformers/fnet/test_modeling.py

robinbg committed:
Merge branch 'robinbg-add-FNet-config' of https://github.com/robinbg/PaddleNLP into robinbg-add-FNet-config

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py
	modified:   tests/transformers/fnet/test_modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Changes to be committed:
	modified:   tests/transformers/fnet/test_modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
	modified:   tests/transformers/fnet/test_modeling.py
	modified:   tests/transformers/test_modeling_common.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
	modified:   tests/transformers/fnet/test_modeling.py
	modified:   tests/transformers/test_modeling_common.py

robinbg committed:
Please enter the commit message for your changes. Lines starting
 On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
Merge branch 'robinbg-add-FNet-config' of https://github.com/robinbg/PaddleNLP into robinbg-add-FNet-config

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/modeling.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/tokenizer.py

robinbg committed:
Update tokenizer.py

robinbg committed:
Update tokenizer.py

robinbg committed:
Update tokenizer.py

robinbg committed:
On branch robinbg-add-FNet-config
 Your branch is up to date with 'origin/robinbg-add-FNet-config'.
 Changes to be committed:
	modified:   paddlenlp/transformers/fnet/tokenizer.py

robinbg commented:
您好，请问这个common test过不了是否和这个错误有关，以及请问能否给些建议， 谢谢。
`sentencepiece_processor.cc(922) LOG(ERROR) src/sentencepiece_processor.cc(289) [model_] Model is not initialized.
[47](https://github.com/PaddlePaddle/PaddleNLP/actions/runs/4279515282/jobs/7450301973#step:5:48)
Returns default value 0`

sijunhe commented on tests/transformers/fnet/test_tokenizer.py:
```diff
@@ -0,0 +1,445 @@
+SAMPLE_VOCAB = get_tests_dir("fixtures/spiece.model")
+
+
+class FNetTokenizationTest(TokenizerTesterMixin, unittest.TestCase):
```
@wj-Mcat 帮忙看一下

sijunhe commented on .copyright.hook:
```diff
@@ -131,4 +131,4 @@ def main(argv=None):
 
 if __name__ == '__main__':
-    exit(main())
+    exit(main())
```
这个修改请revert

wj-Mcat replied:
这里一般是换行的区别，修改起来应该挺简单的。

sijunhe commented on tests/transformers/fnet/test_modeling.py:
```diff
@@ -0,0 +1,306 @@
+
+
+# Override ConfigTester
+class FNetConfigTester(ConfigTester):
```
这不不需要，可以删了

robinbg committed:
On branch robinbg-add-FNet-config

wj-Mcat commented on paddlenlp/transformers/fnet/configuration.py:
```diff
@@ -0,0 +1,141 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```

wj-Mcat commented on paddlenlp/transformers/fnet/modeling.py:
```diff
@@ -116,9 +115,9 @@ def forward(self, hidden_states, output_hidden_states=False, return_dict=True):
 
 class FNetPooler(Layer):
-    def __init__(self, hidden_size):
+    def __init__(self, config):
```
```suggestion
    def __init__(self, config: FNetConfig):
```

在 config 参数上面，建议加上其类型注释，这样在 IDE 当中也会有对应的类型智能提示，从而提升框架的编程体验和代码阅读性。

当然，这个文件中其它相关的构造函数中也建议添加相关的 type hint。

wj-Mcat commented on paddlenlp/transformers/fnet/modeling.py:
```diff
@@ -555,7 +480,7 @@ def forward(
         )
 
-        sequence_output = encoder_outputs[0]
+        sequence_output = encoder_outputs[0] if not return_dict else encoder_outputs["last_hidden_state"]
```
```suggestion
        sequence_output = encoder_outputs[0] if not return_dict else encoder_outputs.last_hidden_state
```

这里可以直接通过属性获取的方式来取值就行了，其它地方也需要注意。

wj-Mcat commented on paddlenlp/transformers/fnet/modeling.py:
```diff
@@ -668,7 +593,7 @@ def forward(
             return_dict=return_dict,
         )
-        pooled_output = outputs[1]
+        pooled_output = outputs[1] if not return_dict else outputs["pooler_output"]
```
跟上面同样的情况，其它地方也都同步修改一下呗。

wj-Mcat commented on paddlenlp/transformers/fnet/modeling.py:
```diff
@@ -686,16 +611,11 @@ class FNetForPreTraining(FNetPretrainedModel):
 
-    def __init__(self, fnet):
-        super().__init__()
+    def __init__(self, config: FNetConfig):
```
这里添加了类型注释就很好。

wj-Mcat commented on paddlenlp/transformers/fnet/modeling.py:
```diff
@@ -912,7 +827,7 @@ def forward(
         if return_dict:
             return {"seq_relationship_logits": seq_relationship_score, "hidden_states": outputs["all_hidden_states"]}
-        return seq_relationship_score, outputs["all_hidden_states"]
+        return seq_relationship_score
```
如果是`return_dict`为`True`，就需要使用`modeling_output`中的对象来构造返回输出，如果里面的数据结构不符合你的要求，此时就需要自定义Output 数据类型，具体可参考：

https://github.com/PaddlePaddle/PaddleNLP/blob/40180768e96e8a85d92f496f555b9b964933c473/paddlenlp/transformers/ernie_ctm/modeling.py#L49-L80

wj-Mcat commented on tests/transformers/fnet/test_modeling.py:
```diff
@@ -0,0 +1,291 @@
+        model = FNetForSequenceClassification(config)
+        model.eval()
+        result = model(input_ids, token_type_ids=token_type_ids, labels=sequence_labels, return_dict=True)
+        self.parent.assertEqual(result["logits"].shape, [self.batch_size, self.num_labels])
```
不建议使用键值对的方式获取模型的输出（虽然可行）
```suggestion
        self.parent.assertEqual(result.logits.shape, [self.batch_size, self.num_labels])
```

wj-Mcat commented on tests/transformers/fnet/test_modeling.py:
```diff
@@ -0,0 +1,291 @@
+        model = FNetForTokenClassification(config=config)
+        model.eval()
+        result = model(input_ids, token_type_ids=token_type_ids, labels=token_labels, return_dict=True)
+        self.parent.assertEqual(result["logits"].shape, [self.batch_size, self.seq_length, self.num_labels])
```
这里也是一样，其它地方同步调整，不一一评论。

wj-Mcat commented on tests/transformers/fnet/test_tokenizer.py:
```diff
@@ -0,0 +1,445 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
```suggestion
# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```

wj-Mcat reviewed (changes requested):
整体完成的很好，就是在小细节上需要调整。
